### DISCUSSION TITLE
Compiling pycryptodome (python2.7)

### DISCUSSION THREAD

**Discussion by Ircama** [Freetz-NG/freetz-ng#947] [Allgemein] [answered]
Can you please advise on how [pycryptodome](https://pycryptodome.readthedocs.io/en/latest/src/installation.html) can be compiled (Python)?

I was able to read [BLE](https://wikipedia.org/wiki/Bluetooth_Low_Energy) Advertisements through an [Ai-Thinker TB-03F-KIT](https://docs.ai-thinker.com/_media/tb-03f-kit_specification_en.pdf) module connected via USB (/dev/ttyUSB0, Driver=ch341,
it needs a simple firmware and a Python code, I'll document this as soon as it is completed), but I would need to decode cyphered messages from BLE devices (like [Xiaomi Mijia BLE sensors](https://esphome.io/components/sensor/xiaomi_ble.html)).

At the moment we only have pycrypto compiled, which does not support AES.MODE_CCM.

Thanks a lot!

**Comment by fda77** [comment]
You habe to add this llike the other pythong things in /make/libs/python-*

**Comment by Ircama** [comment]
It's really hard for me to understand :-(. I think the configuration should need a directory under [freetz-ng/tree/master/make/pkgs](https://github.com/Freetz-NG/freetz-ng/tree/master/make/pkgs), e.g., *python-pycryptodome*.

Then I need to create:

- the "Config.in" file; this looks easy, like [this](https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/python-pycrypto/Config.in)
- the "python-pycryptodome.mk"; this has to be developed similarly to [python-pycrypto.mk](https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/python-pycrypto/python-pycrypto.mk); I think it should point to https://pypi.org/project/pycryptodome/; and somewhere it should be hooked to the toolchain I think
- in case, a "patches" folder, including all needed corrections to the compilation

Really so complex?

**Comment by fda77** [comment]
You have simplified it, but yes

And test it with
make python-XXXX-source
or make python-XXXX-autofix
or make python-XXXX-recompile

**Comment by Ircama** [comment]
Adding a Python module is too complex, it is unclear how to cross-compile a Python wheel. Trying to understand how it was done for some packages, I checked commits that were made many years ago, nothing recent. Besides the device does not have a compiler (you need to cross-compile into some other place). Also Python is the old 2.7. I give up.

Maybe calling C functions from Python is easier.

While the standard build of the firmware works, if I create a new project I cannot even get the cross-compiler to produce an executable that the device can run.

For instance, I have freetz-ng installed in `~/freetz-ng` (Ubuntu 22.4).

```
cd freetz-ng
mkdir example
cd example
cat hello_world.c
#include <stdio.h>
#include <stdlib.h>
int main(void){
  printf("Hello world!\n");
  return 0;
}
```
Which is the right syntax to compile it so that the executable, when moved to the device, runs there?

Also, how this function (my_function.c) can be compiled to create my_function.so running on the device?

```
int square(int i) {
	return i * i;
}
```

**Comment by fda77** [comment]
I never did this, maybe you find some docs if you search for it

**Comment by Ircama** [comment, marked as answer]
Perhaps I found the answer:

- Compilation (cross compiling):

```
/home/$USER/freetz-ng/toolchain/build/mips_gcc-13.2.0_uClibc-1.0.45-nptl_kernel-4.9/mips-linux-uclibc/bin/mips-linux-uclibc-gcc -march=34kc -mtune=34kc -msoft-float -Ofast -pipe -Wa,--trap -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wl,-I/usr/lib/freetz/ld-uClibc.so.1 -o hello_world hello_world.c

/home/$USER/freetz-ng/toolchain/build/mips_gcc-13.2.0_uClibc-1.0.45-nptl_kernel-4.9/mips-linux-uclibc/bin/mips-linux-uclibc-gcc -march=34kc -mtune=34kc -msoft-float -Ofast -pipe -Wa,--trap -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wl,-I/usr/lib/freetz/ld-uClibc.so.1 -shared -o my_function.so -fPIC my_function.c
```

- Execution (device):

```
./hello_world
Hello world!
```

- Python testing:

```
python
```

```python
import ctypes
so_file = "./my_function.so"
my_library = ctypes.CDLL(so_file)
my_library.square(4)
```

```
16
```

**Comment by Ircama** [comment]
ref. https://github.com/Freetz-NG/freetz-ng/pull/1165 for the fix of this issue.
